### ISSUE TITLE
Segmentation fault with Python stuff, in a method of a struct called from struct

### ISSUE THREAD

**Issue by datdinhquoc** [cython/cython#4561]
I have these files to reproduce the problem, mod2.pxd:
```
cdef public:
    void myapi()

    cppclass myclass:
        void mymethod()
```

mod2.pyx
```
cdef public api:
    void myapi():
        print("I am OK (from myapi)")

cdef cppclass myclass:
    void mymethod():
        # Using anything C is OK
        cdef int i = 0
        i += 1

        # Crash with Python var
        # Segmentation fault
        j = 0
        j += 1

        # Crash with Python function
        # Segmentation fault
        print("I am crashing!")
```

mod1.pxd
```
cdef public api:
    void myfunc()
```

mod1.pyx
```
from mod2 cimport myapi, myclass

cdef public cppclass foo:
    void bar():
        print("I am OK (from foo.bar)")

cdef void myfunc():
    print("I am OK")
    myapi()
    cdef foo f
    f.bar()

    # Stack var: Segmentation fault
    print("Calling myclass.mymethod...")
    cdef myclass t
    t.mymethod()

    # # Pointer: Segmentation fault
    # print("Calling myclass.mymethod...")
    # cdef myclass* t2 = new myclass()
    # t2.mymethod()

cpdef test():
    myfunc()
```

Main Python programme app.py
```
import mod1
mod1.test()
```

It's a segmentation fault which happens in a Cython struct `myclass` when using anything Python. Also in mod2, but the global function `myapi` is OK. From a struct method `foo.bar` but in the same file is also OK.

I'm using Python 3.8.2, and Cython 0.29.26

mod1 and mod2 are translated into C++ with `cython -3 --cplus` command, and compiled with `gcc` to .so files.

This situation seems strange, why are Python stuff not touch-able from mod2 (fails in struct only)?

**Comment by scoder** [comment]
Probably not a Cython issue. Make sure you initialise the Python runtime for C++ threads before you use it, and make sure you initialise the cimported API before you use anything from it.

**Comment by datdinhquoc** [comment]
> Probably not a Cython issue. Make sure you initialise the Python runtime for C++ threads before you use it, and make sure you initialise the cimported API before you use anything from it.

Might be interesting, but those files above are not using threads. Should it be related?

**Comment by scoder** [comment]
Could you remove the public and api modifiers from the functions (or everywhere), to see if that makes it work? You don't need them in your example. Even if you want to use the (public) Cython functions in your C++ code, you don't need the api modifiers.

**Comment by datdinhquoc** [comment]
> Could you remove the public and api modifiers from the functions (or everywhere), to see if that makes it work? You don't need them in your example. Even if you want to use the (public) Cython functions in your C++ code, you don't need the api modifiers.

I tried to removed all those modifiers just now but the result is not different. All the `__Pyx_*` calls in `mymethod` fails, but work in `myapi` function although they are in the same file.

**Comment by scoder** [comment]
> As suggested from https://stackoverflow.com/a/55669343/5581893 I tried to add -DCYTHON_PEP489_MULTI_PHASE_INIT=0 to all gcc commands

That's unrelated. IIUC, you're not embedding Python in a C++ application here (and even then, that would be the wrong thing to do).

**Comment by da-woods** [comment]
I just get link errors when I try to import it:

```
undefined symbol: _ZTV21__pyx_t_4mod2_myclass
```

I'd be a little worried that you've done something odd to make it link and your errors are due to that.

**Comment by scoder** [comment]
> > Could you remove the public and api modifiers from the functions (or everywhere)
> >
> I tried to removed all those modifiers just now but the result is not different.

Thanks for verifying that. Seems like it's a Cython issue after all. Sharing the C++ class across modules seems to generate incorrect code that duplicates the C++ class definition on both sides, but only one side knows about the implementation. Meaning, the C++ class is not actually shared.

**Comment by da-woods** [comment]
Yes - almost certainly that. The line that'll be segfaulting will be:

```
__Pyx_INCREF(__pyx_int_0);
```

(i.e. Cython has created a global constant for the Python int 0). Because you've linked it in an odd way, but never actually imported `mod2` that variable isn't initialized.

**Comment by scoder** [comment]
> you've linked it in an odd way

Maybe as well, but I think the issue really is that Cython doesn't know a way to share C++ classes via .pxd files that are only implemented in a .pyx file. The module that cimports that `cppclass` ends up with a C++ class in which all methods are virtual. That's unhelpful.

That duplication is not a problem for C structs, but C++ classes have methods that the other side needs to know about.

**Comment by da-woods** [comment]
It isn't particularly obvious how to do it. For the `myapi` function it's handled via: 

```
__Pyx_ImportFunction(__pyx_t_1, "myapi", (void (**)(void))&__pyx_f_4mod2_myapi, "void (void)") < 0)
```

but I don't think it's possible to do a similar thing for a C++ class (especially for things like destructors, which are invoked automatically so we can't just write a wrapper function to call them)

**Comment by datdinhquoc** [comment]
> I just get link errors when I try to import it:
> 
> ```
> undefined symbol: _ZTV21__pyx_t_4mod2_myclass
> ```
> 
> I'd be a little worried that you've done something odd to make it link and your errors are due to that.

Uhm, after having 2 files .o, link them both to mod1.so, and also to mod2.so to avoid undefined symbol, I did so.

**Comment by datdinhquoc** [comment]
> Sharing the C++ class across modules seems to generate incorrect code that duplicates the C++ class definition on both sides, but only one side knows about the implementation. Meaning, the C++ class is not actually shared.

This cross-module call may be necessary I guess. I'm migrating a Python project to Cython and there are multiple classes with cross-module calls everywhere.

**Comment by scoder** [comment]
> Uhm, after having 2 files .o, link them both to mod1.so, and also to mod2.so to avoid undefined symbol, I did so.

That's not how Python extension modules work. The normal way to use them is one .so file per module. You can link them together, but then you have to tell Python that there is more than one module in it so that it can properly import them all. That's done with the [inittab](https://docs.python.org/3/c-api/import.html#c.PyImport_AppendInittab) mechanism.

**Comment by datdinhquoc** [comment]
Oh yeah, that inittab. What I guess now is that Cython's cimport does list global functions in inittab but not struct members. That's why `myapi` works smoothly with Python vars+funcs but not `mymethod`.

**Comment by da-woods** [comment]
I have a rough plan for how this might work:

1. `cppclass` member functions (including constructors, destructors, operators, etc.) are actually implemented outside the class in a non-member cdef function that can be imported through the existing `__Pyx_ImportFunction` mechanism.
2. Simple inline wrapper functions are generated both in the original module and in the module that the cppclass class is cimported from. E.g.
    ```
    cdef cppclass C:
       C(int val)
       void f()
    ```
    generates something like
    ```c++
    class C {
      C(int val) { C_init_impl(*this, val); }
      void f() { f_impl(*this, val); }
    };
    ```
    in *every* module that uses it. `C_init_impl` and `f_impl` are only generated in the module the class is defined in.
